src/lib.rs

```diff
@@ -420,10 +416,42 @@ macro_rules! derive_alias {(
                 )}
                 #[allow(unused_imports)]
                 pub(in crate) use [< $MacroName __derive_macro >] as $MacroName;
-            )*
-        }
-    )}
-)}
+            }
+        )}
+    };
+    (pub #[derive($MacroName:ident !)] = #[derive($($derives:tt)*)];) => {
+        $crate::ඞ_with_dollar! {( $_:tt ) => (
+            $crate::ඞ::paste! {
+                // To avoid ambiguities with what the re-export
+                // refers to, let's use a hopefully unused name.
+                //
+                // Indeed, eponymous derive macros in scope such as those
+                // from the prelude would otherwise cause trouble with the
+                // re-export line.
+                #[allow(nonstandard_style)]
+                #[macro_export]
+                macro_rules! [< $MacroName __derive_macro >] {(
+                    $_($item:tt)*
+                ) => (
+                    $crate::ඞ_nested_derive! {
+                        #[derive($($derives)*)]
+                        $_($item)*
+                    }
+                )}
+                #[allow(unused_imports)]
+                pub use [< $MacroName __derive_macro >] as $MacroName;
+            }
+        )}
+    };
+    (#[derive($($name_tt:tt)+)] = #[derive($($derives:tt)*)]; $($tt:tt)+) => {
+        $crate::derive_alias!(#[derive($($name_tt)+)] = #[derive($($derives)*)];);
+        $crate::derive_alias!($($tt)+);
+    };
+    (pub #[derive($($name_tt:tt)+)] = #[derive($($derives:tt)*)]; $($tt:tt)+) => {
+        $crate::derive_alias!(pub #[derive($($name_tt)+)] = #[derive($($derives)*)];);
+        $crate::derive_alias!($($tt)+);
+    };
```

Comment:
To avoid syntax duplication (_e.g._, having to handle doc strings here), let's merge these rules with the parent ones.

For instance, the first rule would become:

```rs
macro_rules! derive_alias {
    (
        #[derive($MacroName:ident !)] = #[derive($($derives:tt)*)];
        $($($rest:tt)+)?
    ) => (
        /* current output for `$MacroName` */

        $(
            $crate::derive_alias! { $($rest)+ }
        )?
    );
```

  - also, as a general rule of thumb, nested macro invocation in trailing position are better done using `{}` rather than `();`, just in case the last one expands to as statement expression (_e.g._, `{ 42 }`), so as to avoid discarding its value (_e.g._, the `42: i32`).

Reply:
I don't think I understand what you mean here, in two specific cases:

When you say "let's merge these rules with the parent ones", what is "parent" referring to in this context? Is it positional, like "the rule above this one", or more like "the rule that calls this rule", or something else entirely?

The `{}` vs `()` comment I'm even less clear on. I'm not quite sure *what* doesn't click here for me, but I'll lay out what *does* make sense and with a bit of luck, you might see whatever it is that I'm not seeing. 🤔

 I'm reasonably sure of the terms here, understanding [Items](https://doc.rust-lang.org/reference/items.html) to be distinct from [Statements and Expressions](https://doc.rust-lang.org/reference/statements-and-expressions.html), and in the context of macros, these map to the `fragment-specifier` bits in the documentation for [Metavariables](https://doc.rust-lang.org/reference/macros-by-example.html#metavariables), right? So we're matching those at the macro entry point (_matcher_, but "matching in the matcher" sounds awkward) and then forwarding that on to another macro in the _transcriber_ section? So the docs on [Forwarding a matched fragment](https://doc.rust-lang.org/reference/macros-by-example.html#forwarding-a-matched-fragment) have this bit that seems relevant:

> When forwarding a matched fragment to another macro-by-example, matchers in the second macro will see an opaque AST of the fragment type. The second macro can’t use literal tokens to match the fragments in the matcher, only a fragment specifier of the same type. The `ident`, `lifetime`, and `tt` fragment types are an exception, and can be matched by literal tokens. 

And I'm somewhat fuzzy on this part, but I think that suggests that if we matched a `tt`, it could be kind of anything, but if we matched something more specific, like a `stmt`, `item`, or `expr`, it can only be forwarded to something that accepts those. So it's anything until we treat it as something, and from then on it's only that thing. And since we want to match all of these things, we're not being specific, and just matching some number of `tt` and forwarding them on blindly, right? Which is why we don't know if we have a `stmt`, `item`, `expr`, or something else.

If I've got that right, then I think I understand the context for why the (potential) problem occurs.

But the differences between `{}` and `()` still aren't clear to me at this point.

Looking up some references to figure it out lead me to [this chapter](https://veykril.github.io/tlborm/decl-macros/macros-methodical.html) in The Little Book of Rust Macros, which includes this sentence that I *think* explains the difference?

> In fact, function-like macros can be invoked with any kind of parentheses as well, but invocations with `{ .. }` and `( ... );`, notice the trailing semicolon, are special in that their expansion will always be parsed as an item.

My brain parses that as `{ .. }` and `( ... );` always parse as an `item`, but `[ ... ]` could be something else (like a `stmt` or `expr`?).

Which seems wrong, because you're suggesting a change between two types of parentheses that I (likely incorrectly) think do the same thing?

So... yeah. I'd like to understand this, I've read the docs (not exhaustively), heck, I've even **used** this stuff a decent amount, and it (on the surface at least) works the way I expect it to.

I know that's a wall of text, and I appreciate that it takes time and effort to wade through that. If you happen to notice where my blind spots are, I'd be grateful for the insight, and if I'm misunderstanding something important, I'd welcome any corrections.

Thanks for your patience!